pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/pipeline/PipelineBreakOperator.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.pinot.query.runtime.blocks.TransferableBlockUtils;
+import org.apache.pinot.query.runtime.operator.MultiStageOperator;
+import org.apache.pinot.query.runtime.plan.OpChainExecutionContext;
+
+
+public class PipelineBreakOperator extends MultiStageOperator {
+  private static final String EXPLAIN_NAME = "PIPELINE_BREAKER";
+  private final Deque<Map.Entry<Integer, Operator<TransferableBlock>>> _workerEntries;
+  private final Map<Integer, List<TransferableBlock>> _resultMap;
+  private final CountDownLatch _workerDoneLatch;
+  private TransferableBlock _errorBlock;
+
+
+  public PipelineBreakOperator(OpChainExecutionContext context,
+      Map<Integer, Operator<TransferableBlock>> pipelineWorkerMap) {
+    super(context);
+    _resultMap = new HashMap<>();
+    _workerEntries = new ArrayDeque<>();
+    _workerEntries.addAll(pipelineWorkerMap.entrySet());
+    _workerDoneLatch = new CountDownLatch(1);
+  }
+
+  public Map<Integer, List<TransferableBlock>> getResult() {
+    try {
+      boolean isWorkerDone =
+          _workerDoneLatch.await(_context.getDeadlineMs() - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+      if (isWorkerDone && _errorBlock == null) {
+        return _resultMap;
+      }
+    } catch (Exception e) {
+      _errorBlock = TransferableBlockUtils.getErrorTransferableBlock(e);
+    }
+    return Collections.singletonMap(-1, Collections.singletonList(_errorBlock));
+  }
+
+  @Nullable
+  @Override
+  public String toExplainString() {
+    return EXPLAIN_NAME;
+  }
+
+  @Override
+  protected TransferableBlock getNextBlock() {
+    if (System.currentTimeMillis() > _context.getDeadlineMs()) {
+      _errorBlock = TransferableBlockUtils.getErrorTransferableBlock(QueryException.EXECUTION_TIMEOUT_ERROR);
+      _workerDoneLatch.countDown();
+      return _errorBlock;
+    }
+
+    // Poll from every mailbox operator in round-robin fashion:
+    // - Return the first content block
+    // - If no content block found but there are mailboxes not finished, return no-op block
+    // - If all content blocks are already returned, return end-of-stream block
+    int numWorkers = _workerEntries.size();
+    for (int i = 0; i < numWorkers; i++) {
+      Map.Entry<Integer, Operator<TransferableBlock>> worker = _workerEntries.remove();
+      TransferableBlock block = worker.getValue().nextBlock();
+
+      // Release the mailbox when the block is end-of-stream
+      if (block != null && block.isSuccessfulEndOfStreamBlock()) {
+        continue;
+      }
+
+      // Add the worker back to the queue if the block is not end-of-stream
+      _workerEntries.add(worker);
+      if (block != null) {
+        if (block.isErrorBlock()) {
+          _errorBlock = block;
+          _workerDoneLatch.countDown();
+          return _errorBlock;
+        }
+        List<TransferableBlock> blockList = _resultMap.computeIfAbsent(worker.getKey(), (k) -> new ArrayList<>());
+        blockList.add(block);
```

Comment:
TODO: 
1. only data block should be added, b/c non-datablock are not useful for pipeline breaker results
2. consider a way to pass around stats and metadata, currently these info are discarded.